src/compiler/scanner.ts

```diff
@@ -2469,55 +2468,92 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
-                const saveTokenFlags = tokenFlags;
-                const savePos = pos;
-                const saveEnd = end;
-                pos = tokenStart + 1;
-                end = endOfBody;
-                scanRegularExpressionWorker(regExpFlags, isUnterminated, /*annexB*/ true);
-                tokenStart = saveTokenStart;
-                tokenFlags = saveTokenFlags;
-                pos = savePos;
-                end = saveEnd;
+                    else if (ch === CharacterCodes.closeBracket && characterClassDepth) {
+                        characterClassDepth--;
+                    }
+                    else if (!characterClassDepth) {
+                        if (ch === CharacterCodes.openParen) {
+                            bracketStack.push(CharacterCodes.closeParen);
+                        }
+                        else if (ch === CharacterCodes.openBrace) {
+                            bracketStack.push(CharacterCodes.closeBrace);
+                        }
+                        else if (ch === lastOrUndefined(bracketStack)) {
+                            bracketStack.pop();
+                        }
+                        else if (ch === CharacterCodes.closeParen || ch === CharacterCodes.closeBracket || ch === CharacterCodes.closeBrace) {
+                            // We encountered an unbalanced bracket outside a character class. Treat this position as the end of regex.
+                            break;
+                        }
+                    }
+                    pos++;
+                }
+                // Whitespaces and semicolons at the end are not likely to be part of the regex
+                while (isWhiteSpaceLike(charCodeChecked(pos - 1)) || charCodeChecked(pos - 1) === CharacterCodes.semicolon) pos--;
+                error(Diagnostics.Unterminated_regular_expression_literal, tokenStart, pos - tokenStart);
+            }
+            else {
+                // Consume the slash character
+                pos++;
+                let regExpFlags = RegularExpressionFlags.None;
+                while (pos < end) {
+                    const ch = codePointUnchecked(pos);
+                    if (!isIdentifierPart(ch, languageVersion)) {
+                        break;
+                    }
+                    const size = charSize(ch);
+                    if (reportErrors) {
+                        const flag = characterToRegularExpressionFlag(utf16EncodeAsString(ch));
+                        if (flag === undefined) {
+                            error(Diagnostics.Unknown_regular_expression_flag, pos, size);
+                        }
+                        else if (regExpFlags & flag) {
+                            error(Diagnostics.Duplicate_regular_expression_flag, pos, size);
+                        }
+                        else if (((regExpFlags | flag) & RegularExpressionFlags.UnicodeMode) === RegularExpressionFlags.UnicodeMode) {

```

Comment:
Using `RegularExpressionFlags.UnicodeMode` in this way makes it easy to confuse with `RegularExpressionFlags.Unicode`. I would suggest renaming this flag to `AnyUnicodeMode` to make it less ambiguous.

Reply:
OK (Well, then this should’ve renamed in the spec too. It seems like the old parameter names (U, N) are even less confusing 😅)

Reply:
The change was did in https://github.com/tc39/ecma262/pull/2411, just for my note

Reply:
I think you may be running afoul of the same confusion I'm talking about. `RegularExpressionFlags.Unicode` is synonymous with the `Unicode` (old `U`) production parameter in the spec, while `RegularExpressionFlags.UnicodeSets` is synonymous with the `UnicodeSets` production parameter in the spec.

`RegularExpressionFlags.UnicodeMode` has more in common with the [HasEitherUnicodeFlag](https://tc39.es/ecma262/#sec-runtime-semantics-haseitherunicodeflag-abstract-operation) AO, which is why I'm suggesting we change the name of the `UnicodeMode` flag.

Reply:
Yes, but the names in `RegularExpressionFlags` don't currently match the names in `Pattern`.

```ts
export const enum RegularExpressionFlags {
    None        = 0,
    HasIndices  = 1 << 0, // d
    Global      = 1 << 1, // g
    IgnoreCase  = 1 << 2, // i
    Multiline   = 1 << 3, // m
    DotAll      = 1 << 4, // s
    Unicode     = 1 << 5, // u
    UnicodeSets = 1 << 6, // v
    Sticky      = 1 << 7, // y
    UnicodeMode = Unicode | UnicodeSets,
    Modifiers   = IgnoreCase | Multiline | DotAll,
}
```

Reply:
`RegularExpressionFlags.Unicode` is not synonymous to the `UnicodeMode` parameter. `RegularExpressionFlags.UnicodeMode` is.
![image](https://github.com/microsoft/TypeScript/assets/53967830/751b611d-c584-408d-bce1-5163c4154b68)